File: reference/src/CanonicalTrace.cpp
#include "CanonicalTrace.hpp"

#include <cerrno>
#include <cstring>

namespace
{
constexpr u8 TRACE_MAGIC[8] = {'Z', 'K', 'T', 'H', '0', '6', 'C', 'T'};
constexpr char SCHEMA_DESCRIPTOR[] =
    "zkTH06 canonical trace schema 0.1\n"
    "wire=little-endian;float=ieee754-binary32-raw-bits;coverage=selected-fields\n"
    "subsystems=global,rng,player,player-bullets,enemies-ecl,enemy-bullets,lasers,items,stage,gui-message,effects\n"
    "subsystem-digest=sha256(zkTH06-state-v0.1\\0||subsystem-u16-le||payload)\n"
    "record-root=sha256(zkTH06-trace-root-v0.1\\0||record-prefix||subsystem-records)\n";
constexpr char STATE_DOMAIN[] = "zkTH06-state-v0.1\0";
constexpr char ROOT_DOMAIN[] = "zkTH06-trace-root-v0.1\0";

void SetError(char *error, size_t errorSize, const char *message)
{
    if (error != NULL && errorSize != 0)
    {
        std::snprintf(error, errorSize, "%s", message);
    }
}

class WireBuffer
{
  public:
    WireBuffer(u8 *bytes, size_t capacity) : bytes(bytes), capacity(capacity)
    {
    }

    bool U8(u8 value)
    {
        return this->Append(&value, 1);
    }

    bool U16(u16 value)
    {
        const u8 encoded[2] = {static_cast<u8>(value), static_cast<u8>(value >> 8)};
        return this->Append(encoded, sizeof(encoded));
    }

    bool U32(u32 value)
    {
        const u8 encoded[4] = {static_cast<u8>(value), static_cast<u8>(value >> 8), static_cast<u8>(value >> 16),
                               static_cast<u8>(value >> 24)};
        return this->Append(encoded, sizeof(encoded));
    }

    bool I32(i32 value)
    {
        return this->U32(static_cast<u32>(value));
    }

    bool U64(u64 value)
    {
        u8 encoded[8];
        for (size_t index = 0; index < sizeof(encoded); index++)
        {
            encoded[index] = static_cast<u8>(value >> (index * 8));
        }
        return this->Append(encoded, sizeof(encoded));
    }

    bool Bytes(const void *data, size_t size)
    {
        return this->Append(static_cast<const u8 *>(data), size);
    }

    size_t Size() const
    {
        return this->size;
    }

  private:
    bool Append(const u8 *data, size_t count)
    {
        if (count > this->capacity - this->size)
        {
            return false;
        }
        std::memcpy(this->bytes + this->size, data, count);
        this->size += count;
        return true;
    }

    u8 *bytes;
    size_t capacity;
    size_t size = 0;
};

bool WriteAll(FILE *file, const void *data, size_t size, char *error, size_t errorSize)
{
    if (file == NULL)
    {
        SetError(error, errorSize, "canonical trace file is null");
        return false;
    }
    if (std::fwrite(data, 1, size, file) != size)
    {
        SetError(error, errorSize, "failed to write canonical trace");
        return false;
    }
    return true;
}

CanonicalSubsystem ExpectedSubsystem(size_t index)
{
    return static_cast<CanonicalSubsystem>(index + 1);
}
} // namespace

CanonicalSink::CanonicalSink(CanonicalSubsystem subsystem) : subsystem(subsystem)
{
    this->hash.Update(STATE_DOMAIN, sizeof(STATE_DOMAIN) - 1);
    const u16 id = static_cast<u16>(subsystem);
    const u8 encoded[2] = {static_cast<u8>(id), static_cast<u8>(id >> 8)};
    this->hash.Update(encoded, sizeof(encoded));
}

void CanonicalSink::Payload(const u8 *bytes, size_t size)
{
    this->hash.Update(bytes, size);
    this->byteCount += size;
}

void CanonicalSink::U8(u8 value)
{
    this->Payload(&value, 1);
}

void CanonicalSink::I8(i8 value)
{
    this->U8(static_cast<u8>(value));
}

void CanonicalSink::U16(u16 value)
{
    const u8 encoded[2] = {static_cast<u8>(value), static_cast<u8>(value >> 8)};
    this->Payload(encoded, sizeof(encoded));
}

void CanonicalSink::I16(i16 value)
{
    this->U16(static_cast<u16>(value));
}

void CanonicalSink::U32(u32 value)
{
    const u8 encoded[4] = {static_cast<u8>(value), static_cast<u8>(value >> 8), static_cast<u8>(value >> 16),
                           static_cast<u8>(value >> 24)};
    this->Payload(encoded, sizeof(encoded));
}

void CanonicalSink::I32(i32 value)
{
    this->U32(static_cast<u32>(value));
}

void CanonicalSink::U64(u64 value)
{
    u8 encoded[8];
    for (size_t index = 0; index < sizeof(encoded); index++)
    {
        encoded[index] = static_cast<u8>(value >> (index * 8));
    }
    this->Payload(encoded, sizeof(encoded));
}

void CanonicalSink::F32(f32 value)
{
    u32 bits;
    static_assert(sizeof(bits) == sizeof(value));
    std::memcpy(&bits, &value, sizeof(bits));
    this->U32(bits);
}

void CanonicalSink::Boolean(bool value)
{
    this->U8(value ? 1 : 0);
}

void CanonicalSink::Bytes(const void *data, size_t size)
{
    this->Payload(static_cast<const u8 *>(data), size);
}

CanonicalSubsystemDigest CanonicalSink::Finish(u16 flags, u32 entityCount)
{
    CanonicalSubsystemDigest result;
    result.subsystem = this->subsystem;
    result.flags = flags;
    result.entityCount = entityCount;
    result.byteCount = this->byteCount;
    result.digest = this->hash.Finish();
    return result;
}

Sha256Digest CanonicalTrace::SchemaDigest()
{
    return Sha256::Hash(SCHEMA_DESCRIPTOR, sizeof(SCHEMA_DESCRIPTOR) - 1);
}

const char *CanonicalTrace::SubsystemName(CanonicalSubsystem subsystem)
{
    static constexpr const char *NAMES[CANONICAL_SUBSYSTEM_COUNT] = {
        "global",        "rng",           "player", "player-bullets", "enemies-ecl", "enemy-bullets",
        "lasers",        "items",         "stage",  "gui-message",    "effects",
    };
    const u16 id = static_cast<u16>(subsystem);
    return id >= 1 && id <= CANONICAL_SUBSYSTEM_COUNT ? NAMES[id - 1] : "unknown";
}

CanonicalTerminalReason CanonicalTrace::ParseTerminalReason(const char *reason)
{
    if (reason == NULL)
    {
        return CanonicalTerminalReason::NONE;
    }
    struct Mapping
    {
        const char *name;
        CanonicalTerminalReason value;
    };
    static constexpr Mapping MAPPINGS[] = {
        {"input-error", CanonicalTerminalReason::INPUT_ERROR},
        {"physical-hit", CanonicalTerminalReason::PHYSICAL_HIT},
        {"replay-complete", CanonicalTerminalReason::REPLAY_COMPLETE},
        {"chain-exit-success", CanonicalTerminalReason::CHAIN_EXIT_SUCCESS},
        {"chain-exit-error", CanonicalTerminalReason::CHAIN_EXIT_ERROR},
        {"tick-limit", CanonicalTerminalReason::TICK_LIMIT},
    };
    for (const Mapping &mapping : MAPPINGS)
    {
        if (std::strcmp(reason, mapping.name) == 0)
        {
            return mapping.value;
        }
    }
    return CanonicalTerminalReason::UNKNOWN;
}

bool CanonicalTrace::WriteHeader(FILE *file, const CanonicalRunConfig &config, char *error, size_t errorSize)
{
    std::array<u8, CANONICAL_TRACE_HEADER_SIZE> bytes{};
    WireBuffer wire(bytes.data(), bytes.size());
    const Sha256Digest schemaDigest = CanonicalTrace::SchemaDigest();
    const bool encoded = wire.Bytes(TRACE_MAGIC, sizeof(TRACE_MAGIC)) && wire.U16(VERSION_MAJOR) &&
                         wire.U16(VERSION_MINOR) && wire.U32(CANONICAL_TRACE_HEADER_SIZE) &&
                         wire.U32(CANONICAL_TRACE_RECORD_SIZE) && wire.U16(CANONICAL_SUBSYSTEM_COUNT) &&
                         wire.U16(HEADER_FLAG_SELECTED_FIELDS) && wire.U16(config.initialSeed) &&
                         wire.U8(config.difficulty) && wire.U8(config.character) && wire.U8(config.shotType) &&
                         wire.U8(config.startStage) && wire.U8(static_cast<u8>(config.mode)) && wire.U8(0) &&
                         wire.Bytes(schemaDigest.data(), schemaDigest.size());
    if (!encoded || wire.Size() != bytes.size())
    {
        SetError(error, errorSize, "internal canonical header size mismatch");
        return false;
    }
    return WriteAll(file, bytes.data(), bytes.size(), error, errorSize);
}

bool CanonicalTrace::WriteRecord(FILE *file, const CanonicalFrameMetadata &frame,
                                 const CanonicalSubsystemDigests &subsystems, char *error, size_t errorSize)
{
    std::array<u8, CANONICAL_TRACE_RECORD_SIZE> bytes{};
    WireBuffer wire(bytes.data(), bytes.size());
    bool encoded = wire.U64(frame.tick) && wire.U32(frame.gameFrame) && wire.I32(frame.stage) &&
                   wire.U16(frame.input) && wire.U8(static_cast<u8>(frame.terminalReason)) && wire.U8(frame.flags) &&
                   wire.I32(frame.supervisorState) && wire.U64(frame.recordIndex);
    for (size_t index = 0; encoded && index < subsystems.size(); index++)
    {
        const CanonicalSubsystemDigest &subsystem = subsystems[index];
        if (subsystem.subsystem != ExpectedSubsystem(index))
        {
            SetError(error, errorSize, "canonical subsystem order mismatch");
            return false;
        }
        encoded = wire.U16(static_cast<u16>(subsystem.subsystem)) && wire.U16(subsystem.flags) &&
                  wire.U32(subsystem.entityCount) && wire.U64(subsystem.byteCount) &&
                  wire.Bytes(subsystem.digest.data(), subsystem.digest.size());
    }
    if (!encoded || wire.Size() != CANONICAL_TRACE_RECORD_SIZE - 32)
    {
        SetError(error, errorSize, "internal canonical record size mismatch");
        return false;
    }

    Sha256 root;
    root.Update(ROOT_DOMAIN, sizeof(ROOT_DOMAIN) - 1);
    root.Update(bytes.data(), wire.Size());
    const Sha256Digest rootDigest = root.Finish();
    encoded = wire.Bytes(rootDigest.data(), rootDigest.size());
    if (!encoded || wire.Size() != bytes.size())
    {
        SetError(error, errorSize, "internal canonical root size mismatch");
        return false;
    }
    return WriteAll(file, bytes.data(), bytes.size(), error, errorSize);
}

bool CanonicalTrace::WriteTestFixture(const char *path, char *error, size_t errorSize)
{
    FILE *file = std::fopen(path, "wb");
    if (file == NULL)
    {
        if (error != NULL && errorSize != 0)
        {
            std::snprintf(error, errorSize, "cannot open canonical fixture: %s", std::strerror(errno));
        }
        return false;
    }

    CanonicalRunConfig config;
    config.initialSeed = 0x1234;
    config.difficulty = 2;
    config.character = 1;
    config.shotType = 0;
    config.startStage = 4;
    config.mode = CanonicalRunMode::REPLAY;

    CanonicalSubsystemDigests subsystems;
    for (size_t index = 0; index < subsystems.size(); index++)
    {
        const CanonicalSubsystem id = ExpectedSubsystem(index);
        CanonicalSink sink(id);
        sink.U32(static_cast<u32>(0x10203040 + index));
        sink.F32(static_cast<f32>(index) + 0.5f);
        subsystems[index] = sink.Finish(SUBSYSTEM_FLAG_SELECTED_FIELDS, static_cast<u32>(index + 1));
    }

    CanonicalFrameMetadata frame;
    frame.tick = 0x0102030405060708;
    frame.gameFrame = 12345;
    frame.stage = 4;
    frame.input = 0x55aa;
    frame.terminalReason = CanonicalTerminalReason::TICK_LIMIT;
    frame.flags = 3;
    frame.supervisorState = 7;
    frame.recordIndex = 0;

    bool result = CanonicalTrace::WriteHeader(file, config, error, errorSize) &&
                  CanonicalTrace::WriteRecord(file, frame, subsystems, error, errorSize);
    if (std::fclose(file) != 0 && result)
    {
        SetError(error, errorSize, "failed to close canonical fixture");
        result = false;
    }
    return result;
}

bool CanonicalTrace::SelfTest()
{
    if (!Sha256::SelfTest() || CANONICAL_TRACE_RECORD_SIZE != 592)
    {
        return false;
    }

    CanonicalSink first(CanonicalSubsystem::GLOBAL);
    first.U8(1);
    first.U16(0x0302);
    first.U32(0x07060504);
    first.F32(1.0f);
    const CanonicalSubsystemDigest digest1 = first.Finish(SUBSYSTEM_FLAG_SELECTED_FIELDS, 1);

    const u8 expectedPayload[] = {1, 2, 3, 4, 5, 6, 7, 0, 0, 0x80, 0x3f};
    CanonicalSink second(CanonicalSubsystem::GLOBAL);
    second.Bytes(expectedPayload, sizeof(expectedPayload));
    const CanonicalSubsystemDigest digest2 = second.Finish(SUBSYSTEM_FLAG_SELECTED_FIELDS, 1);
    return digest1.byteCount == sizeof(expectedPayload) && digest1.digest == digest2.digest;
}

File: tools/canonical_trace.py
#!/usr/bin/env python3
"""Decode, validate, and compare zkTH06 canonical digest traces."""

from __future__ import annotations

import argparse
import hashlib
import json
import struct
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import BinaryIO, Iterator


MAGIC = b"ZKTH06CT"
VERSION = (0, 2)
HEADER_SIZE = 64
RECORD_PREFIX_SIZE = 32
SUBSYSTEM_RECORD_SIZE = 48
SUBSYSTEM_NAMES = (
    "global",
    "rng",
    "player",
    "player-bullets",
    "enemies-ecl",
    "enemy-bullets",
    "lasers",
    "items",
    "stage",
    "gui-message",
    "effects",
)
SUBSYSTEM_COUNT = len(SUBSYSTEM_NAMES)
RECORD_SIZE = RECORD_PREFIX_SIZE + SUBSYSTEM_COUNT * SUBSYSTEM_RECORD_SIZE + 32
HEADER_FLAG_SELECTED_FIELDS = 1
SUBSYSTEM_FLAG_SELECTED_FIELDS = 1
SCHEMA_DESCRIPTOR = (
    b"zkTH06 canonical trace schema 0.2\n"
    b"wire=little-endian;float=ieee754-binary32-raw-bits;coverage=selected-fields\n"
    b"projection=runtime-selected-gameplay-v0.2;stable-slots=true;relative-script-offsets=true;anm-future-live=true\n"
    b"subsystems=global,rng,player,player-bullets,enemies-ecl,enemy-bullets,lasers,items,stage,gui-message,effects\n"
    b"subsystem-digest=sha256(zkTH06-state-v0.2\\0||subsystem-u16-le||payload)\n"
    b"record-root=sha256(zkTH06-trace-root-v0.2\\0||record-prefix||subsystem-records)\n"
)
SCHEMA_DIGEST = hashlib.sha256(SCHEMA_DESCRIPTOR).digest()
ROOT_DOMAIN = b"zkTH06-trace-root-v0.2\0"

HEADER_STRUCT = struct.Struct("<8sHHIIHHH6B32s")
RECORD_PREFIX_STRUCT = struct.Struct("<QIiHBBiQ")
SUBSYSTEM_STRUCT = struct.Struct("<HHIQ32s")

RUN_MODES = {0: "unknown", 1: "practice", 2: "replay"}
TERMINAL_REASONS = {
    0: None,
    1: "input-error",
    2: "physical-hit",
    3: "replay-complete",
    4: "chain-exit-success",
    5: "chain-exit-error",
    6: "tick-limit",
    255: "unknown",
}


class CanonicalTraceError(ValueError):
    """Raised when a canonical trace violates its wire-format contract."""


@dataclass(frozen=True)
class TraceHeader:
    initial_seed: int
    difficulty: int
    character: int
    shot_type: int
    start_stage: int
    run_mode: int
    flags: int
    schema_digest: bytes

    def as_dict(self) -> dict[str, object]:
        return {
            "version": f"{VERSION[0]}.{VERSION[1]}",
            "initial_seed": self.initial_seed,
            "difficulty": self.difficulty,
            "character": self.character,
            "shot_type": self.shot_type,
            "start_stage": self.start_stage,
            "run_mode": RUN_MODES[self.run_mode],
            "coverage": "selected-fields",
            "schema_sha256": self.schema_digest.hex(),
            "record_size": RECORD_SIZE,
            "subsystems": list(SUBSYSTEM_NAMES),
        }


@dataclass(frozen=True)
class SubsystemDigest:
    subsystem_id: int
    flags: int
    entity_count: int
    byte_count: int
    digest: bytes

    @property
    def name(self) -> str:
        return SUBSYSTEM_NAMES[self.subsystem_id - 1]

    def as_dict(self) -> dict[str, object]:
        return {
            "id": self.subsystem_id,
            "name": self.name,
            "coverage": "selected-fields",
            "entity_count": self.entity_count,
            "byte_count": self.byte_count,
            "sha256": self.digest.hex(),
        }


@dataclass(frozen=True)
class FrameRecord:
    tick: int
    game_frame: int
    stage: int
    input_mask: int
    terminal_reason: int
    flags: int
    supervisor_state: int
    record_index: int
    subsystems: tuple[SubsystemDigest, ...]
    root_digest: bytes

    def as_dict(self, include_subsystems: bool = True) -> dict[str, object]:
        result: dict[str, object] = {
            "record_index": self.record_index,
            "tick": self.tick,
            "game_frame": self.game_frame,
            "stage": self.stage,
            "input": self.input_mask,
            "terminal_reason": TERMINAL_REASONS[self.terminal_reason],
            "flags": self.flags,
            "supervisor_state": self.supervisor_state,
            "root_sha256": self.root_digest.hex(),
        }
        if include_subsystems:
            result["subsystems"] = [subsystem.as_dict() for subsystem in self.subsystems]
        return result


def _read_exact(file: BinaryIO, size: int, what: str) -> bytes:
    data = file.read(size)
    if len(data) != size:
        raise CanonicalTraceError(f"truncated {what}: expected {size} bytes, got {len(data)}")
    return data


def read_header(file: BinaryIO) -> TraceHeader:
    raw = _read_exact(file, HEADER_SIZE, "canonical trace header")
    (
        magic,
        major,
        minor,
        header_size,
        record_size,
        subsystem_count,
        flags,
        initial_seed,
        difficulty,
        character,
        shot_type,
        start_stage,
        run_mode,
        reserved,
        schema_digest,
    ) = HEADER_STRUCT.unpack(raw)
    if magic != MAGIC:
        raise CanonicalTraceError(f"invalid canonical trace magic: {magic!r}")
    if (major, minor) != VERSION:
        raise CanonicalTraceError(f"unsupported canonical trace version: {major}.{minor}")
    if header_size != HEADER_SIZE:
        raise CanonicalTraceError(f"unexpected canonical header size: {header_size}")
    if record_size != RECORD_SIZE:
        raise CanonicalTraceError(f"unexpected canonical record size: {record_size}")
    if subsystem_count != SUBSYSTEM_COUNT:
        raise CanonicalTraceError(f"unexpected canonical subsystem count: {subsystem_count}")
    if flags != HEADER_FLAG_SELECTED_FIELDS:
        raise CanonicalTraceError(f"unsupported canonical header flags: 0x{flags:04x}")
    if reserved != 0:
        raise CanonicalTraceError("canonical header reserved byte is nonzero")
    if run_mode not in RUN_MODES:
        raise CanonicalTraceError(f"unsupported canonical run mode: {run_mode}")
    if schema_digest != SCHEMA_DIGEST:
        raise CanonicalTraceError(
            f"canonical schema mismatch: expected {SCHEMA_DIGEST.hex()}, got {schema_digest.hex()}"
        )
    return TraceHeader(
        initial_seed=initial_seed,
        difficulty=difficulty,
        character=character,
        shot_type=shot_type,
        start_stage=start_stage,
        run_mode=run_mode,
        flags=flags,
        schema_digest=schema_digest,
    )


def decode_record(raw: bytes, physical_index: int, verify_root: bool = True) -> FrameRecord:
    if len(raw) != RECORD_SIZE:
        raise CanonicalTraceError(f"record {physical_index} has {len(raw)} bytes, expected {RECORD_SIZE}")
    prefix = RECORD_PREFIX_STRUCT.unpack_from(raw, 0)
    tick, game_frame, stage, input_mask, terminal_reason, flags, supervisor_state, record_index = prefix
    if record_index != physical_index:
        raise CanonicalTraceError(
            f"record index mismatch at physical record {physical_index}: encoded {record_index}"
        )
    if terminal_reason not in TERMINAL_REASONS:
        raise CanonicalTraceError(f"record {physical_index} has unknown terminal code {terminal_reason}")

    subsystems: list[SubsystemDigest] = []
    offset = RECORD_PREFIX_SIZE
    for expected_id in range(1, SUBSYSTEM_COUNT + 1):
        subsystem_id, subsystem_flags, entity_count, byte_count, digest = SUBSYSTEM_STRUCT.unpack_from(raw, offset)
        if subsystem_id != expected_id:
            raise CanonicalTraceError(
                f"record {physical_index} subsystem order mismatch: expected {expected_id}, got {subsystem_id}"
            )
        if subsystem_flags != SUBSYSTEM_FLAG_SELECTED_FIELDS:
            raise CanonicalTraceError(
                f"record {physical_index} subsystem {subsystem_id} has unsupported flags 0x{subsystem_flags:04x}"
            )
        subsystems.append(
            SubsystemDigest(
                subsystem_id=subsystem_id,
                flags=subsystem_flags,
                entity_count=entity_count,
                byte_count=byte_count,
                digest=digest,
            )
        )
        offset += SUBSYSTEM_RECORD_SIZE

    root_digest = raw[-32:]
    if verify_root:
        expected_root = hashlib.sha256(ROOT_DOMAIN + raw[:-32]).digest()
        if root_digest != expected_root:
            raise CanonicalTraceError(
                f"record {physical_index} root mismatch: expected {expected_root.hex()}, got {root_digest.hex()}"
            )
    return FrameRecord(
        tick=tick,
        game_frame=game_frame,
        stage=stage,
        input_mask=input_mask,
        terminal_reason=terminal_reason,
        flags=flags,
        supervisor_state=supervisor_state,
        record_index=record_index,
        subsystems=tuple(subsystems),
        root_digest=root_digest,
    )


def iter_records(file: BinaryIO, verify_root: bool = True) -> Iterator[FrameRecord]:
    physical_index = 0
    while True:
        raw = file.read(RECORD_SIZE)
        if not raw:
            return
        if len(raw) != RECORD_SIZE:
            raise CanonicalTraceError(
                f"truncated canonical record {physical_index}: expected {RECORD_SIZE} bytes, got {len(raw)}"
            )
        yield decode_record(raw, physical_index, verify_root=verify_root)
        physical_index += 1


def read_trace(path: Path, verify_root: bool = True) -> tuple[TraceHeader, list[FrameRecord]]:
    with path.open("rb") as file:
        header = read_header(file)
        records = list(iter_records(file, verify_root=verify_root))
    return header, records


def summarize(path: Path) -> dict[str, object]:
    with path.open("rb") as file:
        header = read_header(file)
        first: FrameRecord | None = None
        last: FrameRecord | None = None
        count = 0
        total_payload_bytes = 0
        subsystem_stats = [
            {"total_bytes": 0, "min_bytes": None, "max_bytes": 0, "max_entities": 0}
            for _ in SUBSYSTEM_NAMES
        ]
        for record in iter_records(file):
            if first is None:
                first = record
            last = record
            count += 1
            for index, subsystem in enumerate(record.subsystems):
                stats = subsystem_stats[index]
                stats["total_bytes"] += subsystem.byte_count
                stats["min_bytes"] = (
                    subsystem.byte_count
                    if stats["min_bytes"] is None
                    else min(stats["min_bytes"], subsystem.byte_count)
                )
                stats["max_bytes"] = max(stats["max_bytes"], subsystem.byte_count)
                stats["max_entities"] = max(stats["max_entities"], subsystem.entity_count)
                total_payload_bytes += subsystem.byte_count
    named_stats = {
        name: {
            **stats,
            "min_bytes": 0 if stats["min_bytes"] is None else stats["min_bytes"],
        }
        for name, stats in zip(SUBSYSTEM_NAMES, subsystem_stats, strict=True)
    }
    return {
        "valid": True,
        "path": str(path),
        "size": path.stat().st_size,
        "header": header.as_dict(),
        "record_count": count,
        "hashed_payload_bytes": total_payload_bytes,
        "subsystem_stats": named_stats,
        "first_record": None if first is None else first.as_dict(include_subsystems=True),
        "last_record": None if last is None else last.as_dict(include_subsystems=True),
    }


def compare(left_path: Path, right_path: Path) -> dict[str, object]:
    with left_path.open("rb") as left_file, right_path.open("rb") as right_file:
        left_header = read_header(left_file)
        right_header = read_header(right_file)
        if left_header != right_header:
            return {
                "equal": False,
                "kind": "header",
                "left": left_header.as_dict(),
                "right": right_header.as_dict(),
            }

        left_records = iter_records(left_file)
        right_records = iter_records(right_file)
        index = 0
        while True:
            left = next(left_records, None)
            right = next(right_records, None)
            if left is None or right is None:
                if left is None and right is None:
                    return {"equal": True, "record_count": index}
                return {
                    "equal": False,
                    "kind": "length",
                    "matching_records": index,
                    "left_ended": left is None,
                    "right_ended": right is None,
                }
            if left.root_digest != right.root_digest:
                changed_subsystems = [
                    left_subsystem.name
                    for left_subsystem, right_subsystem in zip(left.subsystems, right.subsystems, strict=True)
                    if left_subsystem != right_subsystem
                ]
                return {
                    "equal": False,
                    "kind": "record",
                    "first_mismatch": index,
                    "changed_subsystems": changed_subsystems,
                    "left": left.as_dict(include_subsystems=True),
                    "right": right.as_dict(include_subsystems=True),
                }
            index += 1


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("trace", type=Path, help="canonical trace to validate")
    parser.add_argument("--compare", type=Path, metavar="OTHER", help="find the first mismatch against OTHER")
    args = parser.parse_args()
    try:
        result = compare(args.trace, args.compare) if args.compare is not None else summarize(args.trace)
    except (CanonicalTraceError, OSError) as error:
        print(json.dumps({"valid": False, "error": str(error)}, indent=2), file=sys.stderr)
        return 2
    print(json.dumps(result, indent=2))
    return 0 if args.compare is None or result["equal"] else 1


if __name__ == "__main__":
    raise SystemExit(main())
